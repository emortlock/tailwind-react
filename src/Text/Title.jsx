import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { TailwindContext } from '../context'

const Title = ({
  children,
  className,
  is,
  size,
  subtitle,
  flush,
  h,
  ...rest
}) => {
  const level = h || Math.max(7 - size, 1)
  const Component = is || `h${level}`
  let ariaProps = {}

  if (!subtitle && Component !== 'string' && !/h[1-6]/i.test(Component)) {
    ariaProps = {
      role: 'heading',
      'aria-level': level,
    }
  }
  return (
    <TailwindContext.Consumer>
      {theme => (
        <Component
          {...rest}
          {...ariaProps}
          className={classnames(
            `font-${theme.text.family[subtitle ? 'subtitle' : 'title']}`,
            'leading-tight',
            `text-${theme.text.size.title[size - 1]}`,
            !subtitle && `text-${theme.textColors.emphasis} font-bold`,
            subtitle && `text-${theme.textColors.body} font-medium`,
            !flush && `mb-${theme.spacing.md}`,
            className,
          )}
        >
          {children}
        </Component>
      )}
    </TailwindContext.Consumer>
  )
}

Title.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  h: PropTypes.number,
}

Title.defaultProps = {
  children: undefined,
  className: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  h: undefined,
}

export default Title
