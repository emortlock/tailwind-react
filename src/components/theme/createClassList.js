import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
