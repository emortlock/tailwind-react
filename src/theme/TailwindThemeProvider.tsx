import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
