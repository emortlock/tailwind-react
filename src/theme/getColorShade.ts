const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
