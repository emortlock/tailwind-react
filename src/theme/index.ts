export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
