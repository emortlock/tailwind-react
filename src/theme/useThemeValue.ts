import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
