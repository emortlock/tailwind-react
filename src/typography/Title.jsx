import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const Title = ({
  theme,
  children,
  className,
  is,
  size,
  subtitle,
  flush,
  h,
  ...rest
}) => {
  const level = h || Math.max(7 - size, 1)
  const Component = is || `h${level}`
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  let ariaProps = {}

  if (!subtitle && Component !== 'string' && !/h[1-6]/i.test(Component)) {
    ariaProps = {
      role: 'heading',
      'aria-level': level,
    }
  }
  return (
    <Component
      {...filterProps(rest, tailwindProps)}
      {...ariaProps}
      className={classnames(
        `font-${theme.text.family[subtitle ? 'subtitle' : 'title']}`,
        'leading-tight',
        `text-${theme.text.size.title[size - 1]}`,
        !subtitle && `text-${theme.textColors.emphasis} font-bold`,
        subtitle && `text-${theme.textColors.body} font-medium`,
        !flush && `mb-${theme.spacing.md}`,
        userClassNames,
      )}
    >
      {children}
    </Component>
  )
}

Title.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  h: PropTypes.number,
  ...tailwindPropTypes,
}

Title.defaultProps = {
  children: undefined,
  className: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  h: undefined,
}

export { Title as component }
export default withTheme(Title)
