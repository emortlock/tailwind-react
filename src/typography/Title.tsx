import type { TextProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'

import { RawText as Text } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface TitleProps<E extends HTMLElement = HTMLHeadingElement>
  extends WithTheme,
    Omit<TextProps<E>, 'size'> {
  subtitle?: boolean
  flush?: boolean
  level?: number
  size?: number
}

export const RawTitle = <E extends HTMLElement = HTMLHeadingElement>({
  theme,
  children,
  as = 'h1',
  size = 4,
  subtitle,
  flush,
  level = 1,
  ...rest
}: TitleProps<E>) => {
  const hLevel = level || Math.max(7 - size, 1)
  const element: TitleProps['as'] = as || `h${hLevel}`

  let ariaProps: TextProps<E> = {}

  if (!subtitle && typeof element === 'string' && !/h[1-6]/i.test(element)) {
    ariaProps = {
      'role': 'heading',
      'aria-level': hLevel,
    }
  }

  return (
    <Text<E>
      as={element}
      {...ariaProps}
      leading="tight"
      font={[
        theme.text.family[subtitle ? 'subtitle' : 'title'],
        subtitle ? 'medium' : 'bold',
      ]}
      text={[
        theme.text.size.title[size - 1],
        subtitle ? theme.textColors.body : theme.textColors.emphasas,
      ]}
      m={!flush ? { b: theme.spacing.md } : undefined}
      {...rest}
    >
      {children}
    </Text>
  )
}

export const Title = withTheme<TitleProps>(RawTitle)
