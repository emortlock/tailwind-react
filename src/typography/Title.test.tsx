import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawTitle as Title } from './Title'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = render(<Title {...props}>Title</Title>)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('Title', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup({ size: 3 })

    expect(asFragment()).toMatchSnapshot()
  })
})
