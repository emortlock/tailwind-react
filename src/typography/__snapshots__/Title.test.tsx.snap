// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title renders matching snapshot 1`] = `
<DocumentFragment>
  <h1
    class="font-sans font-bold text-2xl text-black leading-tight mb-4"
  >
    Title
  </h1>
</DocumentFragment>
`;
